### DISCUSSION TITLE
In app router, server component, how can I get the current url (similar of window.location)

### DISCUSSION THREAD

**Discussion by GuyAvraham** [vercel/next.js#63587] [App Router] [answered]
Im using App router.
and have a place in my server component where I need to know the requested URL. 
If I was in a client component, I could use window.location.origin.

GPT offered me to try 
import { useServerRequest } from 'next/navigation';
but looks like this is not a real thing. 
any suggestions?

**Comment by aziyatali** [comment, marked as answer]
`request.nextUrl.pathname;`

For more, [here](https://nextjs.org/docs/app/building-your-application/routing/middleware)

**Comment by GuyAvraham** [comment]
It is not a rest api handler, so I don't have a request. 
the context is inside a server component, and the only props I have are "params" and "searchParams"

**Reply by GuyAvraham** [reply]
oh, Ok, now I understand:
using middleware I can save the request.nexturl.pathname on the headers, and then in the server component I can access the headers to read the URL. 
thanks.

**Reply by 2333Ge** [reply]
excuse me, I dont understand，how can I get the pathname in async page 😭，can you give me an example

**Reply by Set27** [reply]
@2333Ge https://www.propelauth.com/post/getting-url-in-next-server-components

**Reply by Venipa** [reply]
> @2333Ge https://www.propelauth.com/post/getting-url-in-next-server-components

please dont use this... worst example

**Reply by 2333Ge** [reply]
@Venipa  why not?

**Reply by Venipa** [reply]
> @Venipa why not?

I may have overreacted there a bit but I just don't understand why nextjs provide a context to pages instead we need to workaround this issue

**Reply by 2333Ge** [reply]
> > @Venipa why not?
> 
> I may have overreacted there a bit but I just don't understand why nextjs provide a context to pages instead we need to workaround this issue

Agree, it should be provided by Nextjs. For now, we at least have a temporary solution

**Reply by Fedec96** [reply]
@Set27 That is indeed an interesting approach, but it's breaking my use case: it triggers when I fire the server action that allows next-intl to switch locale. Anyone has a nice middleware matcher config that properly ignores statics, API routes, server actions and public content?

**Comment by tusharjain3105** [comment]
You can get it from headers

**Comment by DanielSalazarLoaiza** [comment]
useServerRequest doesn’t exist in Next.js.
If you’re in a Server Component and need to know the requested URL, you can use the new headers()
 function from next/headers.

Example:

```
import { headers } from 'next/headers';

`export` default function MyComponent() {
  const h = headers();
  const host = h.get('host');
  const protocol = h.get('x-forwarded-proto') || 'http';
  const fullUrl = `${protocol}://${host}`;

  return <div>Requested URL: {fullUrl}</div>;
}
```


✅ Works in Server Components
✅ Doesn’t rely on window
✅ Safe for use during SSR or Edge rendering

If you also need the full path (not just origin), you can read:

const path = h.get('x-invoke-path') || '';
const fullUrl = `${protocol}://${host}${path}`;

**Comment by Ridd0** [comment]
It's a shame we have to get creative to make it work. I wanted to add `view` param, my code:

In your middleware add:
`response.headers.set('x-query-string', request.nextUrl.search);`

and in your server component:

```
const headersList = await headers();

// Extract searchParams from current URL query string (set by middleware)
let searchParams: { view?: string } = {};
const queryString = headersList.get('x-query-string');

if (queryString) {
  try {
    const params = new URLSearchParams(queryString);
    const view = params.get('view');
    // Validate: only allow 'XYZ' as valid value for security
    if (view === 'XYZ') {
      searchParams = { view };
    }
  } catch (e) {
    // Ignore parsing errors
  }
}
```

**Reply by mthanhtran** [reply]
I feel this is security concern since headers is not safe.
